pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1023 @@
+
+		let storage_bond_in_user_currency =
+			Self::convert_to_currency_if_asset_is_not_native(sold_currency, storage_bond_in_native_currency)?;
+
+		Ok(storage_bond_in_user_currency)
+	}
+
+	fn get_on_initialize_weight() -> u64 {
+		crate::weights::HydraWeight::<T>::on_initialize().ref_time()
+	}
+
+	fn get_execute_schedule_weight() -> u64 {
+		crate::weights::HydraWeight::<T>::execute_schedule().ref_time()
+	}
+
+	fn take_transaction_fee_from_user(owner: &T::AccountId, order: &Order<<T as Config>::Asset>) -> DispatchResult {
+		let fee_currency = Self::get_asset_in(order);
+
+		let fee_amount_in_sold_asset = Self::get_transaction_fee(fee_currency)?;
+
+		T::Currency::transfer(
+			fee_currency.into(),
+			owner,
+			&T::FeeReceiver::get(),
+			fee_amount_in_sold_asset.into(),
+		)?;
+
+		Ok(())
+	}
+
+	fn reserve_named_reserve(
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+		who: &T::AccountId,
+	) -> DispatchResult {
+		let currency_for_reserve = match schedule.order {
+			Order::Buy { asset_in, .. } => asset_in,
+			Order::Sell { asset_in, .. } => asset_in,
+		};
+
+		T::Currency::reserve_named(
+			&NAMED_RESERVE_ID,
+			currency_for_reserve.into(),
+			who,
+			schedule.total_amount.into(),
+		)?;
+
+		Ok(())
+	}
+
+	fn unreserve_all_named_reserved_sold_currency(schedule_id: ScheduleId, who: &T::AccountId) -> DispatchResult {
+		let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotExist)?;
+		let sold_currency = Self::get_asset_in(&schedule.order);
+		T::Currency::unreserve_all_named(&NAMED_RESERVE_ID, sold_currency.into(), who);
+
+		Ok(())
+	}
+
+	fn get_asset_in(order: &Order<T::Asset>) -> <T as Config>::Asset {
+		let sold_currency = match order {
+			Order::Sell { asset_in, .. } => asset_in,
+			Order::Buy { asset_in, .. } => asset_in,
+		};
+		*sold_currency
+	}
+
+	fn get_asset_out(order: &Order<T::Asset>) -> <T as Config>::Asset {
+		let asset_out = match order {
+			Order::Sell { asset_out, .. } => asset_out,
+			Order::Buy { asset_out, .. } => asset_out,
+		};
+		*asset_out
+	}
+
+	fn weight_to_fee(weight: Weight) -> Balance {
+		// cap the weight to the maximum defined in runtime, otherwise it will be the
+		// `Bounded` maximum of its data type, which is not desired.
+		let capped_weight: Weight = weight.min(T::BlockWeights::get().max_block);
+		<T as pallet::Config>::WeightToFee::weight_to_fee(&capped_weight)
+	}
+
+	fn add_schedule_id_to_existing_ids_per_block(
+		next_schedule_id: ScheduleId,
+		blocknumber_for_schedule: <T as frame_system::Config>::BlockNumber,
+	) -> DispatchResult {
+		let schedule_ids =
+			ScheduleIdsPerBlock::<T>::get(blocknumber_for_schedule).ok_or(Error::<T>::NoScheduleIdsPlannedInBlock)?;
+		if schedule_ids.len() == T::MaxSchedulePerBlock::get() as usize {
+			let mut consequent_block = blocknumber_for_schedule;
+			consequent_block.saturating_inc();
+			Self::plan_schedule_for_block(consequent_block, next_schedule_id)?;
```

Comment:
this is an indirect recursive call that could lead to an arbitrary amount of storage reads if all the schedules are full.
have you modeled that attack?

Reply:
I could come up with something like this:
```
	fn plan_schedule_for_block(b: T::BlockNumber, schedule_id: ScheduleId) -> DispatchResult {
		let mut blocknumber_for_schedule = b;

		while ScheduleIdsPerBlock::<T>::contains_key(blocknumber_for_schedule) {
			let schedule_ids = ScheduleIdsPerBlock::<T>::get(blocknumber_for_schedule);
			if schedule_ids.len() < T::MaxSchedulePerBlock::get() as usize {
				Self::add_schedule_id_to_block(schedule_id, blocknumber_for_schedule)?;
				return Ok(());
			}
			blocknumber_for_schedule.saturating_inc();
		}

		let vec_with_first_schedule_id = Self::create_bounded_vec(schedule_id)?;
		ScheduleIdsPerBlock::<T>::insert(blocknumber_for_schedule, vec_with_first_schedule_id);

		Ok(())
	}

	fn add_schedule_id_to_block(
		next_schedule_id: ScheduleId,
		blocknumber_for_schedule: T::BlockNumber,
	) -> DispatchResult {
		ScheduleIdsPerBlock::<T>::try_mutate_exists(blocknumber_for_schedule, |schedule_ids| -> DispatchResult {
			let schedule_ids = schedule_ids.as_mut().ok_or(Error::<T>::NoScheduleIdsPlannedInBlock)?;

			schedule_ids
				.try_push(next_schedule_id)
				.map_err(|_| Error::<T>::InvalidState)?;
			Ok(())
		})?;

		Ok(())
	}
	```
	
	There is no recursion now, only the while loop goes as long as it finds a block where we can plan the schedule.
	
	What do you think of it?